### PR TITLE
typed: Allow duplicates when we merge

### PR DESCRIPTION
**Pull request by apelisse**
The goal is to ignore duplicates if they are not included in the partial object, and to replace all of the occurences if they are in the partial object.

/assign @liggitt 
/cc @alexzielenski

### CODE REVIEW

File: typed/merge_test.go [hunk 1]
```diff
@@ -423,6 +467,18 @@ var mergeCases = []mergeTestCase{{
 		`{"atomicList":["a","a","a"]}`,
 		`{"atomicList":["a","a"]}`,
 		`{"atomicList":["a","a"]}`,
+	}, {
+		`{"list":[{"key":"a","id":1,"bv":true},{"key":"b","id":2},{"key":"a","id":1,"bv":false,"nv":2}]}`,
+		`{"list":[{"key":"a","id":1,"nv":3}, {"key":"c","id":3},{"key":"b","id":2}]}`,
+		`{"list":[{"key":"a","id":1,"bv":true,"nv":3},{"key":"c","id":3},{"key":"b","id":2}]}`,
```
Comment (apelisse) [inline typed/merge_test.go:473]: So this is merging the new item with the first occurence, I'm currently thinking we should replace altogether.

File: typed/merge.go [hunk 2]
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
...
 				continue
 			}
 		}
 		if lI < lLen {
-			pe := lhsOrder[lI]
+			pe := lhsPEs[lI]
 			if _, ok := observedRHS.Get(pe); !ok {
 				// take LHS item
-				lChild, _ := observedLHS.Get(pe)
+				lChild := lhs.AtUsing(w.allocator, lI)
```
Comment (alexzielenski) [inline typed/merge.go:241]: This seems equivalent, since we are looking LHS first. Just curious why the change? Is it a noop?

File: typed/merge.go [hunk 3]
```diff
@@ -290,11 +300,9 @@ func (w *mergingWalker) indexListPathElements(t *schema.List, list value.List) (
 			// this element.
 			continue
 		}
-		if _, found := observed.Get(pe); found {
-			errs = append(errs, errorf("duplicate entries for key %v", pe.String())...)
-			continue
+		if _, found := observed.Get(pe); !found {
+			observed.Insert(pe, child)
```
Comment (alexzielenski) [inline typed/merge.go:304]: Is this existence check necessary? Is `observed` a set?

File: typed/merge.go [hunk 4]
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
...
+				lChild := lhs.AtUsing(w.allocator, lI)
 				mergeOut, errs := w.mergeListItem(t, pe, lChild, nil)
 				errs = append(errs, errs...)
 				if mergeOut != nil {
 					out = append(out, *mergeOut)
 				}
 				lI++
 				continue
+			} else if _, ok := mergedRHS.Get(pe); ok {
+				// we've already merged it with RHS, we don't want to duplicate it, skip it.
```
Comment (alexzielenski) [inline typed/merge.go:250]: I'm not sure why this case is necessary. It's hard to track the invariants of `observedRHS` vs `mergedRHS` without going very deep into the code.

File: typed/merge.go [hunk 5]
```diff
@@ -180,14 +180,18 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 	}
 	out := make([]interface{}, 0, outLen)
 
-	rhsOrder, observedRHS, rhsErrs := w.indexListPathElements(t, rhs)
+	rhsPEs, observedRHS, rhsErrs := w.indexListPathElements(t, rhs)
```
Comment (alexzielenski) [inline typed/merge.go:183]: Are the PE's no longer ordered?

File: typed/merge_test.go (hunk 1 above)
Comment (alexzielenski) [reply typed/merge_test.go:473]: This case surprised me. From our prior conversations I would not have expected the merge with one of the duplicates. It doesn't seem like as the author of such a resource I'd be very conscious of which of the duplicates I am merging with.


Is it possible to have two instances of the same value but they are owned by different field managers? In that case I almost would expect "MY" version of the duplicate to be merged; but don't think that'd be worth the effort.

**Review by alexzielenski — COMMENTED** [review]
Except for one tests make a lot of sense and very straightforward.

TBH I could not gain deep sense of the code changes without obtaining degree in SMD. This code could benefit a lot from some more inline comments about the invariants of this algorithm.

File: typed/merge.go (hunk 5 above)
Comment (apelisse) [reply typed/merge.go:183]: They are still, but "order" didn't really tell me what they were. I could revert or make it `rhsOrderedPEs` or something.

File: typed/merge.go (hunk 2 above)
Comment (apelisse) [reply typed/merge.go:241]: Actually, lhs can have multiple times the same PE, so we want to make sure we get the right one (I don't think it's equivalent).

File: typed/merge.go (hunk 3 above)
Comment (apelisse) [reply typed/merge.go:304]: observed is a map from pathelement to the actual value.

the value that ends-up being in the PE is the one we're going to use to merge I think, so if we were to override here, I think we would end-up merging with the last item in the list rather than the first. But since we've decided we might not want to do that, I will probably deal with that differently.

File: typed/merge_test.go (hunk 1 above)
Comment (apelisse) [reply typed/merge_test.go:473]: > Is it possible to have two instances of the same value but they are owned by different field managers?

No.

> This case surprised me. From our prior conversations I would not have expected the merge with one of the duplicates. It doesn't seem like as the author of such a resource I'd be very conscious of which of the duplicates I am merging with.

yeah, I think we just need to replace rather than merge, which means I'll need to know if they are duplicated and replace altogether, I'll update the algorithm.

File: typed/merge.go [hunk 6]
```diff
@@ -180,14 +180,18 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 	}
```
Comment (liggitt) [inline typed/merge.go:180]: AI: look for issue reported where someone couldn't update because of managed fields

File: typed/merge.go [hunk 7]
```diff
@@ -199,12 +203,14 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
...
+	lLen, rLen = len(lhsPEs), len(rhsPEs)
 	for lI, rI := 0, 0; lI < lLen || rI < rLen; {
 		if lI < lLen && rI < rLen {
-			pe := lhsOrder[lI]
-			if pe.Equals(rhsOrder[rI]) {
+			pe := lhsPEs[lI]
+			if pe.Equals(rhsPEs[rI]) {
 				// merge LHS & RHS items
+				mergedRHS.Insert(pe, struct{}{})
 				lChild, _ := observedLHS.Get(pe)
```
Comment (liggitt) [inline typed/merge.go:214]: comment that `lChild` will be nil if the LHS had duplicates for pe

File: typed/merge.go [hunk 8]
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
...
 				// shared item, but not the one we want in this round
 				lI++
 				continue
 			}
 		}
 		if lI < lLen {
-			pe := lhsOrder[lI]
+			pe := lhsPEs[lI]
 			if _, ok := observedRHS.Get(pe); !ok {
 				// take LHS item
```
Comment (liggitt) [inline typed/merge.go:240]: comment that we use lhs.AtUsing instead of observedLHS because observedLHS can contain nil if the lhs had duplicates

File: typed/merge.go [hunk 9]
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
...
+				// we've already merged it with RHS, we don't want to duplicate it, skip it.
+				lI++
 			}
 		}
 		if rI < rLen {
 			// Take the RHS item, merge with matching LHS item if possible
-			pe := rhsOrder[rI]
+			pe := rhsPEs[rI]
+			mergedRHS.Insert(pe, struct{}{})
 			lChild, _ := observedLHS.Get(pe) // may be nil
```
Comment (liggitt) [inline typed/merge.go:258]:
```suggestion
			lChild, _ := observedLHS.Get(pe) // may be nil if not present or LHS duplicated this item
```

File: typed/merge.go [hunk 10]
```diff
@@ -291,10 +301,10 @@ func (w *mergingWalker) indexListPathElements(t *schema.List, list value.List) (
 			continue
 		}
 		if _, found := observed.Get(pe); found {
-			errs = append(errs, errorf("duplicate entries for key %v", pe.String())...)
-			continue
+			observed.Insert(pe, value.NewValueInterface(nil))
```
Comment (liggitt) [inline typed/merge.go:304]: is it possible to *only* tolerate duplicates making the LHS and keep returning the duplicate error making the RHS?

**Review by liggitt — COMMENTED** [review]
tests look good, the change was shockingly small.

mostly requests for clearer comments, one request to keep the duplicate-error-returning bit the same when constructing observedRHS

File: typed/merge.go (hunk 2 above)
Comment (apelisse) [reply typed/merge.go:241]: Added a comment for that.

**Comment by apelisse** [comment]
PTAL, I think I've addressed all the commits. Also split the rename of rhsOrder and lhsOrder in a seperate commit to keep the meat of the change isolated.

**Comment by liggitt** [comment]
Thanks, open the k/k PR to update master and ping me for review.

I suspect we'll want to backport this, not sure if that means picking the fix to other smd branches or not.
